### PR TITLE
crane / zot launcher: kill registry serve process on shell exit

### LINKED ISSUE #419: `crane registry serve` daemons are not cleaned up
When executing build actions locally, `crane registry serve` processes are cleaned up after a `bazel build`.
However, when using remote execution (in my case with buildbarn), `crane registry serve` processes survive between actions and pile up over time on the remote executor. This results in hundreds of processes and will starve the host of memory over multiple days.

## How to reproduce

- [setup a buildbarn cluster](https://github.com/buildbarn/bb-deployments) (for example using the docker-compose setup)
- Use the buildbarn cluster for remote execution (`--remote_executor=...`)
- use rules_oci to build container images
- Look at the running processes after the build is complete

### Expected behavior

`crane registry serve` should no longer run

### Actual behavior

`crane registry serve` processes are still running

### PR DESCRIPTION
**Pull request by malt3**
Fixes #419

### CODE REVIEW

**Review by thesayyn — COMMENTED** [review]
I believe the cleanup procedure doesn't need to leak into call site rules. the launcher script can easily keep a global state and expose a cleanup function that kills the process.

**Comment by malt3** [comment]
> [...] the launcher script can easily keep a global state and expose a cleanup function that kills the process
This is not correct. The `start_registry` function is invoked in a subshell like this:

```bash
readonly REGISTRY=$(start_registry "${STORAGE_DIR}" "${STDERR}" "${REGISTRY_PIDFILE}")
```

The subshell (and any global state that may be preserved inside) does not survive the subshell invocation.

Or to be more precise:

```bash
function start_registry() {
    # [...]
    "${REGISTRY_BIN}" >> $output 2>&1 &
    readonly REGISTRY_PID=$!
    # [...]
     return 0
}

function stop_registry() {
    if [[ -z "${REGISTRY_PID+x}" ]]; then
        echo "Registry not started"
        return 0
    fi
    echo "Stopping registry process ${REGISTRY_PID}"
    kill -9 "${REGISTRY_PID}" || true
    return 0
}

# if I do this, the variable will survive but we cannot easily capture the registry address without writing to a file:
start_registry # [...] args
stop_registry

# if I do this, the variable is only part of the subshell
readonly REGISTRY=$(start_registry # [...] args)
stop_registry
```

**Comment by thesayyn** [comment]
Ah true, we can write the pid into storage_dir/proc.pid or call `pkill -P $$` kill all subprocesses. I'd rather keep the diff minimal.

**Comment by malt3** [comment]
I think pkill should not be used since it is not part of POSIX and not even in coreutils or util-linux.
I used the fixed path inside the storage dir, as suggested.

File: oci/private/image.sh.tpl [hunk 1]
```diff
@@ -12,6 +12,8 @@ readonly STORAGE_DIR="{{storage_dir}}"
 
 readonly STDERR=$(mktemp)
 
+trap "stop_registry ${STORAGE_DIR}" EXIT
```
Comment (thesayyn) [inline oci/private/image.sh.tpl:15]: this overrides the existing trap on EXIT called `silent_on_success` can you just call the stop registry function there?

File: examples/attest/test.bash [hunk 2]
```diff
@@ -15,21 +15,22 @@ readonly SBOM_PATH="$7"
 # Launch a registry instance at a random port
 source "${REGISTRY_LAUNCHER}"
 REGISTRY=$(start_registry $TEST_TMPDIR $TEST_TMPDIR/output.log)
+trap "stop_registry ${TEST_TMPDIR}" EXIT
```
Comment (thesayyn) [inline examples/attest/test.bash:18]: these don't need to change as they are tests. we can keep them as is as we don't have the ghost problem in our CI.

**Review by thesayyn — COMMENTED** [review]
diff is still big and there seems to be some formatting changes. Can you revert those?

File: oci/private/registry/crane_launcher.sh.tpl [hunk 3]
```diff
@@ -26,4 +28,16 @@ function start_registry() {
...
+}
+
+function stop_registry() {
+    local storage_dir="$1"
+    local registry_pid="$1/proc.pid"
+    if [[ ! -f "${registry_pid}" ]]; then
+        echo "Registry not started"
+        return 0
+    fi
+    echo "Stopping registry process"
```
Comment (thesayyn) [inline oci/private/registry/crane_launcher.sh.tpl:40]: unnecessary noise? did you want to print this to `>&2`?

**Comment by thesayyn** [comment]
Sorry @malt3 going to look into this soon.

**Review by thesayyn — APPROVED** [review]

**Pull request merged**